[WIP] Merge interop spec

MariusVanDerWijden opened the pull request:
This PR contains a first implementation of the interop spec rebased on current master

to test:
```
./build/bin/geth --dev --catalyst --http --ws -http.api "engine" console
```

```json
curl -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"engine_preparePayload","params":[{"parentHash":"0x582ea3ee1731029756c8bd211647e02fb68528a1d0a18a99a413d8b6afbe20b2", "timestamp":"0x1", "random":"0x0000000000000000000000000000000000000000000000000000000000000000", "feeRecipient":"0x0000000000000000000000000000000000000000"}],"id":67}' http://localhost:8545
{"jsonrpc":"2.0","id":67,"result":{"payloadId":"0x0"}}


curl -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"engine_getPayload","params":["0x0"],"id":67}' http://localhost:8545
{"jsonrpc":"2.0","id":67,"result":{"blockHash":"0xca6fc9b73939ad40c0ebfcab5b93ca81548f690b62d3c54b0fd0bdcbb2ce5872","parentHash":"0x582ea3ee1731029756c8bd211647e02fb68528a1d0a18a99a413d8b6afbe20b2","coinbase":"0xcd3f20468170e3094f09f16ae987e6325634ada1","stateRoot":"0x6cb6d72150ab4f942667c2f3426091b7de8095e08c5e8fc5c92edeb2d804729b","receiptsRoot":"0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421","logsBloom":"0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","random":"0x0000000000000000000000000000000000000000000000000000000000000000","number":"0x1","gasLimit":"0xaf79e0","gasUsed":"0x0","timestamp":"0x1","extraData":null,"baseFeePerGas":"0x0","transactions":[]}}



 curl -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"engine_executePayload","params":[{"blockHash":"0xca6fc9b73939ad40c0ebfcab5b93ca81548f690b62d3c54b0fd0bdcbb2ce5872","parentHash":"0x582ea3ee1731029756c8bd211647e02fb68528a1d0a18a99a413d8b6afbe20b2","coinbase":"0xcd3f20468170e3094f09f16ae987e6325634ada1","stateRoot":"0x6cb6d72150ab4f942667c2f3426091b7de8095e08c5e8fc5c92edeb2d804729b","receiptsRoot":"0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421","logsBloom":"0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","random":"0x0000000000000000000000000000000000000000000000000000000000000000","number":"0x1","gasLimit":"0xaf79e0","gasUsed":"0x0","timestamp":"0x1","extraData":"","baseFeePerGas":"0x0","transactions":[]}],"id":67}' http://localhost:8545
{"jsonrpc":"2.0","id":67,"result":{"status":"VALID"}}


 curl -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"engine_consensusValidated","params":[{"blockHash":"0xca6fc9b73939ad40c0ebfcab5b93ca81548f690b62d3c54b0fd0bdcbb2ce5872", "status":"VALID"}],"id":67}' http://localhost:8545
{"jsonrpc":"2.0","id":67,"result":null}


curl -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"engine_forkChoiceUpdated","params":[{"headBlockHash":"0xca6fc9b73939ad40c0ebfcab5b93ca81548f690b62d3c54b0fd0bdcbb2ce5872", "finalizedBlockHash":"0xca6fc9b73939ad40c0ebfcab5b93ca81548f690b62d3c54b0fd0bdcbb2ce5872"}],"id":67}' http://localhost:8545
{"jsonrpc":"2.0","id":67,"result":null}
```

lightclient commented on eth/catalyst/api_types.go:
```diff
@@ -21,50 +21,82 @@ import (
+	ParentHash    common.Hash    `json:"parentHash"    gencodec:"required"`
+	Coinbase      common.Address `json:"coinbase"      gencodec:"required"`
+	StateRoot     common.Hash    `json:"stateRoot"     gencodec:"required"`
+	ReceiptRoot   common.Hash    `json:"receiptsRoot"  gencodec:"required"`
```
```suggestion
	ReceiptRoot   common.Hash    `json:"receiptRoot"  gencodec:"required"`
```
I believe this should be `receiptRoot`, not `receiptsRoot`.

lightclient replied:
https://github.com/ethereum/execution-apis/blob/main/src/engine/interop/specification.md#executionpayload

lightclient commented on eth/catalyst/api_types.go:
```diff
@@ -21,50 +21,82 @@ import (
+	ReceiptRoot   common.Hash    `json:"receiptsRoot"  gencodec:"required"`
+	LogsBloom     []byte         `json:"logsBloom"     gencodec:"required"`
+	Random        common.Hash    `json:"random"        gencodec:"required"`
+	Number        uint64         `json:"number"        gencodec:"required"`
```
```suggestion
	Number        uint64         `json:"blockNumber"        gencodec:"required"`
```

lightclient commented on consensus/beacon/consensus.go:
```diff
@@ -0,0 +1,401 @@
+// (b) the timestamp is not verified anymore
+func (beacon *Beacon) verifyHeader(chain consensus.ChainHeaderReader, header, parent *types.Header) error {
+	// Ensure that the header's extra-data section is of a reasonable size
+	if len(header.Extra) != 0 {
```
Should this still that the header extra is less than `params.MaximumExtraDataSize`?

lightclient replied:
cc @mkalinin

mkalinin replied:
Right, this check must stay the same as it is in the case of PoW blocks, in the geth code it looks as follows:
https://github.com/ethereum/go-ethereum/blob/f2491c5ed77b9554bd979a426db741a99acf6fa4/consensus/ethash/consensus.go#L259-L262

lightclient commented on eth/catalyst/api.go:
```diff
@@ -127,12 +270,10 @@ func (api *consensusAPI) AssembleBlock(params assembleBlockParams) (*executableD
 		return nil, err
 	}
-
 	coinbase, err := api.eth.Etherbase()
```
Should this respect the the `AssembleBlockParams`'s `FeeRecipient`?

lightclient commented on eth/catalyst/api_types.go:
```diff
@@ -21,50 +21,82 @@ import (
+}
+
+// JSON type overrides for payloadResponse.
+type payloadResponseMarshaling struct {
```
This should technically be a `hexutil.Uint64` instead of an object for the interop, but the plan is to make this object after.

lightclient commented on eth/catalyst/api.go:
```diff
@@ -250,66 +381,159 @@ func decodeTransactions(enc [][]byte) ([]*types.Transaction, error) {
 		GasLimit:    params.GasLimit,
 		GasUsed:     params.GasUsed,
 		Time:        params.Timestamp,
+		// TODO (MariusVanDerWijden) add params.Random to header once required
```
I think `Extra` should be here?

MariusVanDerWijden replied:
I thought we're not going to have extradata as a header field anymore?

lightclient replied:
I assumed that it's a defined field in [`ExecutionPayload`](https://github.com/ethereum/execution-apis/blob/main/src/engine/interop/specification.md#executionpayload) and there is no mention of how to deal with it in EIP-3675 that it would continue to exist.

paulhauner commented:
Hi there, thanks for publishing the [test vectors](https://hackmd.io/@tvanepps/amphora-milestones). It's much appreciated!

I have been working on these vectors and have found some discrepancies. I've raised them on TG/Discord, but I thought it might be helpful to have them collated on Github.

Here's the discrepencies:

- For the `ExecutionPayload` object, the test vectors use `receiptsRoot` whilst the spec uses `receiptRoot`.
- Also for `ExecutionPayload`, the test vectors use `number` whilst the spec uses `blockNumber`.
- For `engine_preparePayload`, the test vectors state `{"jsonrpc":"2.0","id":67,"result":{"payloadId":"0x0"}`  whist I understand that the spec declares `{"jsonrpc":"2.0","id":67,"result":"0x0"}` (return a `QUANTITY` not an `Object`).
    - Credit to @tersec for first raising this, who reports that Geth returns as per the spec, not as per the vectors.
- Test test vectors use `engine_forkChoiceUpdated` whilst the spec declares `engine_forkchoiceUpdated` (no capital `C`).

MariusVanDerWijden commented:
Thank you for looking into it! Most of the issues were already fixed in code. 
I've now updated the test vectors to be in line with the implementation.
And I've updated `engine_forkChoiceUpdated` -> `engine_forkchoiceUpdated`

paulhauner commented:
I just ran the Lighthouse tests against Geth and I can confirm all those previous issues have been fixed. Thank you :tada: 

There are still two things remaining:

- Geth doesn't return an `Object` for the `engine_preparePayload` as per the [`v1.0.0.alpha.2`](https://github.com/ethereum/execution-apis/blob/v1.0.0-alpha.2/src/engine/interop/specification.md) changes. That's totally understandable, it was only released hours ago. I thought I'd just make a note of it.
- Geth doesn't accept `"finalizedBlockHash":"0x0000000000000000000000000000000000000000000000000000000000000000"` on `engine_forkchoiceUpdated`. (More below)

According to [EIP-3675](https://eips.ethereum.org/EIPS/eip-3675#definitions):

> A POS_FORKCHOICE_UPDATED event contains references to the head of the canonical chain and to the most recent finalized block. Before the first finalized block occurs in the system the finalized block hash provided by this event is stubbed with 0x0000000000000000000000000000000000000000000000000000000000000000.

This issue can be recreated with a modified version of the test vectors:

Request:
```
curl -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","method":"engine_forkchoiceUpdated","params":[{"headBlockHash":"0xb084c10440f05f5a23a55d1d7ebcb1b3892935fb56f23cdc9a7f42c348eed174", "finalizedBlockHash":"0x0000000000000000000000000000000000000000000000000000000000000000"}],"id":67}' 
```
Response:
```
http://localhost:8545 {"jsonrpc":"2.0","id":67,"error":{"code":4,"message":"unknown header"}}
```

terencechain commented on eth/catalyst/api_types.go:
```diff
@@ -21,50 +21,82 @@ import (
+	GasUsed       uint64         `json:"gasUsed"       gencodec:"required"`
+	Timestamp     uint64         `json:"timestamp"     gencodec:"required"`
+	ExtraData     []byte         `json:"extraData"     gencodec:"required"`
+	BaseFeePerGas uint64         `json:"baseFeePerGas" gencodec:"required"`
```
Is `uint64` correct? `baseFeePerGas` is defined as `QUANTITY, 256 Bits` in the latest engine API spec

holiman commented on eth/catalyst/api.go:
```diff
@@ -105,9 +173,105 @@ func (api *consensusAPI) makeEnv(parent *types.Block, header *types.Header) (*bl
+	}
+	parent := api.eth.BlockChain().GetBlockByHash(params.ParentHash)
+	if parent == nil {
+		return INVALID, fmt.Errorf("could not find parent %x", params.ParentHash)
```
I don't think this will work. Our RPC server implementation, AFAIK, when dealing with a non-nil return and a non-nil error, will discard the return vaue and return only the error. This might even be a 'feature' of json rpc: either you return a response XOR an error. Our server chooses the error.

jwasinger commented on core/blockchain.go:
```diff
@@ -205,15 +203,17 @@ type BlockChain struct {
 
-	shouldPreserve func(*types.Block) bool // Function used to determine whether should preserve the given block.
+	shouldPreserve  func(*types.Block) bool        // Function used to determine whether should preserve the given block.
+	terminateInsert func(common.Hash, uint64) bool // Testing hook used to terminate ancient receipt chain insertion.
```
stale code?  I don't see any references to this elsewhere.

blazejkrzak commented on consensus/beacon/consensus.go:
```diff
@@ -0,0 +1,401 @@
+// (b) the timestamp is not verified anymore
+func (beacon *Beacon) verifyHeader(chain consensus.ChainHeaderReader, header, parent *types.Header) error {
+	// Ensure that the header's extra-data section is of a reasonable size
+	if len(header.Extra) > 32 {
```
`32` should go to enum or const

blazejkrzak commented on consensus/beacon/consensus.go:
```diff
@@ -0,0 +1,401 @@
+		return fmt.Errorf("invalid difficulty: have %v, want %v", header.Difficulty, beaconDifficulty)
+	}
+	// Verify that the gas limit is <= 2^63-1
+	cap := uint64(0x7fffffffffffffff)
```
same here, it could be extracted outside the function
